containers/Connection.ts

```diff
@@ -1,54 +1,112 @@
 import { createContainer } from "unstated-next";
 import { useState, useEffect } from "react";
 import { ethers } from "ethers";
+import { API as OnboardApi, Wallet } from "bnc-onboard/dist/src/interfaces";
+import { API as NotifyApi } from "bnc-notify/dist/src/interfaces";
+import Onboard from "bnc-onboard";
+import Notify from "bnc-notify";
 import { Observable } from "rxjs";
 import { debounceTime } from "rxjs/operators";
 
 type Provider = ethers.providers.Provider;
 type Block = ethers.providers.Block;
 type Network = ethers.providers.Network;
-type ExternalProvider = ethers.providers.ExternalProvider;
 type Signer = ethers.Signer;
 
 function useConnection() {
   const [provider, setProvider] = useState<Provider | null>(null);
+  const [onboard, setOnboard] = useState<OnboardApi | null>(null);
+  const [notify, setNotify] = useState<NotifyApi | null>(null);
   const [signer, setSigner] = useState<Signer | null>(null);
   const [network, setNetwork] = useState<Network | null>(null);
   const [address, setAddress] = useState<string | null>(null);
+  const [selectedWallet, setSelectedWallet] = useState<string | null>(null);
   const [error, setError] = useState<Error | null>(null);
   const [block$, setBlock$] = useState<Observable<Block> | null>(null);
 
   const attemptConnection = async () => {
-    if (window.ethereum === undefined) {
-      throw Error("MetaMask not found, please visit https://metamask.io/");
-    }
-
-    // // for testing
-    // const provider = new ethers.providers.JsonRpcProvider()
-    // let signer = new ethers.Wallet(process.env.PRIV_KEY)
-    // signer = signer.connect(provider)
+    const apiKey = process.env.REACT_APP_ONBOARD_API_KEY
+      ? process.env.REACT_APP_ONBOARD_API_KEY
+      : "12153f55-f29e-4f11-aa07-90f10da5d778";
+    const infuraId =
+      process.env.REACT_APP_INFURA_ID || "d5e29c9b9a9d4116a7348113f57770a8";
+    const infuraRpc = `https://${network?.name}.infura.io/v3/${infuraId}`;
+    const onboard = Onboard({
+      dappId: apiKey,
+      hideBranding: true,
+      networkId: 1, // Default to main net. If on a different network will change with the subscription.
+      subscriptions: {
+        address: setAddress,
+        network: async (networkId: any) => {
+          onboard.config({ networkId: networkId });
+        },
+        wallet: async (wallet: Wallet) => {
+          if (wallet.provider) {
+            const ethersProvider = new ethers.providers.Web3Provider(
+              wallet.provider
+            );
+            setProvider(ethersProvider);
+            setNetwork(await ethersProvider.getNetwork());
+            setSigner(ethersProvider.getSigner());
+            await ethersProvider.send("eth_requestAccounts", []);
+            setAddress(await ethersProvider.getSigner().getAddress());
+            setSelectedWallet(wallet.name);
+          } else {
```

Comment:
I think you should set the Provider and all of these values to null together. Is it possible for network, signer, address, etc. to be non null if provider `wallet.provider` doesn't exist?

Reply:
done with setting everything else to null.

I'm not sure about how it handles edge cases. need to test that a bit more.